### ISSUE TITLE
Cross-reference equations, figures and tables when bookdown is NOT already installed?

### ISSUE THREAD

**Issue by andrewjohnlowe** [rstudio/bookdown#512]
This question is identical to [this question](https://stackoverflow.com/questions/47794311/knitr-cross-reference-equations-figures-and-tables-when-bookdown-is-not-alread) that I posted on StackOverflow last month, but for which I didn't get a satisfactory answer. Apologies for the duplication. I'm posting here because I suspect the most elegant solution might be to handle this issue in RStudio, as I will explain later. I hope this might also kick-start some activity in this direction.

I'm aware that the canonical method for cross-referencing figures and tables in a PDF produced by knitr is to use `bookdown::pdf_document2`.

*This is awesome!* But suppose that I want to send my Rmd to a user who doesn't have **bookdown** already installed? How can I ensure that the document will render if knitted in an environment in which **bookdown** is not already installed? I want to have everything self-contained in a single Rmd file. The consumer of the document might be someone with little or no prior experience of R, and has just installed R and RStudio but no packages. How can I ensure that if they click **knit** that they will get the PDF with cross-references? Alternatively, what if Binder or similar provide the means to execute (or, better still, knit to PDF) Rmd files in much the same way as is done with Jupyter Notebooks -- I'd like to be able to bundle everything that is needed for rendering the PDF in an automated way in the Rmd file itself.

An example document looks like this:

    ---
    title: "Cross-refs test"
    author: "Me"
    date: "13 December 2017"
    output:
      bookdown::pdf_document2:
        citation_package: natbib
        fig_caption: yes
        keep_tex: yes
    ---

    ```{r setup, include=FALSE}
    knitr::opts_chunk$set(echo = TRUE)
    ```

    ## Title

    \begin{equation}
    f\left(k\right)=\binom{n}{k}p^k\left(1-p\right)^{n-k} \label{eq:binom}
    \end{equation}

    You may refer to it using `\@ref(eq:binom)`, e.g., see Equation \@ref(eq:binom).

So if this Rmd file is downloaded and knit by a brain-dead zombie, or by Binder, how can I make sure that this doesn't happen:

    Error in loadNamespace(name) : there is no package called 'bookdown'
    Calls: <Anonymous> ... tryCatch -> tryCatchList -> tryCatchOne -> <Anonymous>
    Execution halted

I can't assume that a consumer of my Rmd will know what to do in this situation, nor any automated process that I don't control that renders a PDF from the Rmd that it grabs from a repo somewhere (like what Binder does).

If there are *no* cross-references, I can specify `pdf_document` as the output format and RStudio will read the YAML and if **knitr** and **rmarkdown** are not installed, installation of these will be triggered automatically (the user/brain-dead zombie just has to click on "OK" to accept installation) and the PDF will be rendered. So if these two packages are missing: no problem, at least when using RStudio. User gets Rmd, installs R and Rstudio, knits the Rmd, and the PDF appears. The same is not true if using **bookdown** to get cross-referencing.

It seems the YAML is interpreted before R code is executed, so the same kind of logic as this that is used to print the current time in a document (``` date: "`r Sys.time()`"```) doesn't work. Otherwise, I'd `r install.packages("bookdown")` here.

So, is there a method to do cross-referencing without bookdown, but as simply and as elegantly as with bookdown, or is there someway of triggering knitr to install bookdown such that rendering won't crash when encountering `bookdown::pdf_document2`?

Basically, what I want is cross-references in a knitted PDF with minimal user intervention to make it happen -- 'cos my user might be a brain-dead zombie or robot.

I want to avoid hard-coding cross-refs. And I'd rather not use LaTeX. Just pure R Markdown and some knitr magic, if possible.

A possible solution to this issue would be for RStudio to recognise instances of bookdown::pdf_document2 in the YAML and trigger installation of bookdown, in the same way that the user is prompted to install knitr and rmarkdown if they are not already present: they need only click a button to proceed and complete the process of rendering the document. It would appear that this would be the most elegant solution. If this is not possible, what's the next best solution?

Thanks!

**Comment by yihui** [comment]
This issue is not specific to **bookdown**, but exists for all R packages. It is just an exception that **knitr** and **rmarkdown** could be automatically installed (only in the RStudio IDE).

If your user has no knowledge of R, she has already had to do the following things before she could build PDF:

1. Install R;
2. Install RStudio;
3. Install LaTeX.

I think these things are way more complicated than a single command in the R console: `install.packages('bookdown')`. Therefore I'm not convinced that she would feel so much more painful to do one more thing after installing all these packages.

I don't want to backport the features to **rmarkdown**, because the number of **rmarkdown** users is much bigger than bookdown users, which means I need to be much more careful and conservative in terms of making changes the **rmarkdown** package. If I happen to break anything in it, it is actually quite a pain to update it on CRAN (it can be very painful and stressful to deal with CRAN when you maintain a popular package).

I'll post possible solutions to your problem on SO soon. Thanks for the detailed report!

**Comment by andrewjohnlowe** [comment]
Thanks for the reply. I look forward to hearing your thoughts on a possible solution to my problem.

I accept your point that typing `install.packages('bookdown')` is no more difficult than installing R, RStudio, and LaTeX, but I have a special use-case in mind that I have intimated but that I have not previously stated explicitly:

I'd like users to be able to reproduce an elegantly-formatted document containing an analysis *without* needing to install R, RStudio, or LaTeX! The user would only need to click on a hyperlink on a webpage (in the abstract of the paper posted on bioRxiv or arXiv, for example) or *in the PDF of the paper itself* to launch an executable environment that recreates the exact same paper with all the numeric results and plots. The technology to do this could be based on [Binder](https://mybinder.org/). Getting this to work is like a jigsaw puzzle: I can see enough of the picture to be encouraged that this is feasible, but I'm missing some pieces of the puzzle. **This issue corresponds to one of those missing pieces.** I want to remove the need for human intervention.

Currently it is possible to run Python analyses in Jupyter Notebooks that are hosted in a GitHub repo and launched with Binder. Such as this [this example](https://github.com/jakevdp/PythonDataScienceHandbook). It is also possible to do this with R analyses, such as in [this example](https://github.com/cms-opendata-education/cms-rmaterial-multiple-languages). No installation required. You can even run these on your smartphone. It's also possible to [launch RStudio in Binder](https://github.com/binder-examples/dockerfile-rstudio). Also, files created in Binder are accessible in the Jupyter file list. So, in principal, it should be possible to enable users to click on a link that launches the analysis in Binder and renders the TeX and PDF which they can then download. Not only would these be identical to the journal article TeX and PDF, they could in fact be the ultimate source of the manuscript to be submitted to the journal. Binder and, behind the scenes, Docker provide the means to ensure that all users have the same execution environment. [Packrat](https://cran.r-project.org/web/packages/packrat/index.html), or perhaps the Dockerfile, ensure everyone is using the same package versions and their dependencies. [Zenodo](https://zenodo.org/) can mint a DOI for the GitHub repo for the Binder that can be cited.

I'm still figuring this out. Some parts of this are still fuzzy. I haven't connected all the dots yet. Some of my thinking is captured [here](https://github.com/jupyterhub/binderhub/issues/261).

Basically, what I'm trying to make happen is:

1. User gets PDF of paper
2. User clicks on hyperlink (in PDF?)
3. ?????
4. Analysis runs
5. The same PDF is regenerated -- the analysis is reproducible, Q.E.D.

And I need some method to enable cross-referencing in the paper.

I imagine I could create some special file in my GitHub repo that provides instructions to install bookdown before rendering the PDF with `output: bookdown::pdf_document2`. But wouldn't it be cool if I didn't have to do this? It's already possible to launch a Jupyter Notebook in Binder from a ipynb file in any of the GitHub repos out there, and there's even browser add-ons that let you do this just by clicking a button. Similarly, it would be really cool if I --- or anyone else --- could use Binder to render a PDF from any Rmd in someone else's repo (that doesn't contain special files). There's lots of Rmd files floating around on GitHub! Anyway, that's my dream. Maybe it's a naive fantasy. Sorry for the long post. It ran a bit off-topic too.
